gk/fib.c

```diff
@@ -1361,45 +1361,97 @@ check_gateway_prefix(struct ip_prefix *prefix, struct ipaddr *gw_addr)
 
 /*
  * This function makes sure that only a drop or another grantor entry
- * must be able to be longer than a grantor or a drop prefix.
+ * can have a longer prefix than a grantor or a drop entry.
  *
- * TODO A potential bug needs to be fixed here.
- * Assuming there's prefix 10.1.1.0/24 that leads to a gateway and
- * the prefix being added to a grantor is 10.1/16.
- * That is, the gateway entry would be a security hole.
- *
- * The solution is to extend RTE LPM table to have an iterator over
- * its entries. The iterator should require a prefix as parameter and
- * list all entries as long as a given prefix or longer.
- * To list all entries, the prefix would have length zero.
+ * The importance of this sanity check is illustrated in the following example:
+ * assume that the prefix 10.1.1.0/24 forwards to a gateway and
+ * the prefix 10.1/16 being added forwards to a grantor.
+ * Although the prefix 10.1/16 is intended to protect every host in that
+ * destination, the prefix 10.1.1.0/24 leaves some of those hosts unprotected.
+ * Without this sanity check, variations of this example could go unnoticed
+ * in deployments of Gatekeeper until it is too late.
  */
 static int
 check_prefix_locked(struct ip_prefix *prefix,
 	enum gk_fib_action action, struct gk_config *gk_conf)
 {
 	uint8_t i;
 	int ip_prefix_present;
+	uint32_t prefix_ip4 = ntohl(prefix->addr.ip.v4.s_addr);
```

Comment:
Variable `prefix_ip4` shouldn't be here since `check_prefix_locked()` may be called for an IPv6 prefix.